kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java

```diff
@@ -355,74 +346,150 @@ private Module resolveConfigBubbles(Definition def, Module inputModule, RuleGram
-                    .flatMap(b -> performParse(cache.getCache(), parser, realScanner, b))
-                    .map(this::upClaim)
-                .collect(Collections.toSet());
-
-            Set<Sentence> ruleSet = stream(module.localSentences())
-                    .parallel()
-                    .filter(s -> s instanceof Bubble)
-                    .map(b -> (Bubble) b)
-                    .filter(b -> b.sentenceType().equals(rule))
-                    .flatMap(b -> performParse(cache.getCache(), parser, realScanner, b))
-                    .map(this::upRule)
-                .collect(Collections.toSet());
-
-            Set<Sentence> contextSet = stream(module.localSentences())
-                    .parallel()
-                    .filter(s -> s instanceof Bubble)
-                    .map(b -> (Bubble) b)
-                    .filter(b -> b.sentenceType().equals(context))
-                    .flatMap(b -> performParse(cache.getCache(), parser, realScanner, b))
-                    .map(this::upContext)
-                .collect(Collections.toSet());
-
-            Set<Sentence> aliasSet = stream(module.localSentences())
-                    .parallel()
-                    .filter(s -> s instanceof Bubble)
-                    .map(b -> (Bubble) b)
-                    .filter(b -> b.sentenceType().equals(alias))
-                    .flatMap(b -> performParse(cache.getCache(), parser, realScanner, b))
-                    .map(this::upAlias)
-                .collect(Collections.toSet());
-
-            if (needNewScanner) {
-                realScanner.close();//required for Windows.
+            if (!replacedBubbles.isEmpty()) {
+                // make a new module with replaced bubbles
+                return Module(m.name(),
+                        m.imports(),
+                        stream((Set<Sentence>) m.localSentences().$minus$minus(immutable(replacedBubbles)).
+                                $bar(immutable(cachedParses))).collect(Collections.toSet()),
+                        m.att());
+            }
+            return m;
+        }, "").apply(defWithConfig);
+
+        // prepare scanners for remaining bubbles
+        // scanners can be reused so find the bottom modules which include all other modules
+        java.util.Set<Module> botMods = getBotModules(defWithCaches.modules()).stream().filter(m -> m.sentences().filter(s -> s instanceof Bubble).size() != 0).collect(Collectors.toSet());
+        // prefer modules that import the main module. This way we avoid using the main syntax module which could contain problematic syntax for rule parsing
+        java.util.Set<Module> orderedBotMods = new java.util.LinkedHashSet<>();
+        for (Module m : botMods) {
+            if (m.name().equals(defWithCaches.mainModule().name()) || m.importedModuleNames().contains(defWithCaches.mainModule().name()))
+                orderedBotMods.add(m);
+        }
+        orderedBotMods.addAll(botMods);
```

Comment:
Once again, should we call these `entryMods` instead, to have consistent terminology? `botMods` to me sound like modules that are at the "bottom" of the import chain, meaning they import no other modules.

Reply:
Also, why does `botMods` not include any modules with bubbles?

Reply:
Should call them `topModules` then?
I'm using visual cues when deciding what to call top and what is bottom, because that's how you write a definition. But semantically I guess the main module is a top module.